### PR TITLE
Canny with OpenVX

### PR DESCRIPTION
**Pull request by apavlenko**

### CODE REVIEW

File: modules/core/src/precomp.hpp [hunk 1]
```diff
@@ -258,15 +258,13 @@ struct ImplCollector
 struct CoreTLSData
 {
     CoreTLSData() :
-//#ifdef HAVE_OPENCL
         device(0), useOpenCL(-1),
-//#endif
```
Comment (alalek) [inline modules/core/src/precomp.hpp]: This is TODO mark, no need to remove it.

File: modules/core/include/opencv2/core/ovx_defs.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+CV_EXPORTS_W bool useOpenVX();
+
+/// Enable/disable use of OpenVX
+CV_EXPORTS_W void setUseOpenVX(bool flag);
+} // namespace cv
+
+// utility macro for running OpenVX-based implementations
+#ifdef HAVE_OPENVX
+
+#define CV_OVX_RUN(condition, func, ...)          \
```
Comment (alalek) [inline modules/core/include/opencv2/core/ovx_defs.hpp]: This is not related to public interface. You need to slit this file into public API and private OpenCV-only API.

File: modules/core/src/precomp.hpp (hunk 1 above)
Comment (apavlenko) [inline modules/core/src/precomp.hpp]: @alalek, could you provide more details why these lines should be left?
(they look like a garbage)

File: modules/imgproc/src/canny.cpp [hunk 3]
```diff
@@ -775,6 +777,57 @@ class finalPass : public ParallelLoopBody
...
+    vx_border_mode_t border;
+    border.mode = VX_BORDER_MODE_REPLICATE;
+    const vx_enum VX_NODE_BORDER = VX_NODE_ATTRIBUTE_BORDER_MODE;
+#endif
+    IVX_CHECK_STATUS( vxSetNodeAttribute(node, VX_NODE_BORDER, &border, sizeof(border)) );
+    graph.verify();
+    graph.process();
+
+    //IVX_CHECK_STATUS( vxuCannyEdgeDetector(context, _src, threshold, kSize, (useL2 ? VX_NORM_L2 : VX_NORM_L1), _dst) );
+    }
```
Comment (savuor) [inline modules/imgproc/src/canny.cpp]: do not forget to swap handles in the end

**Comment by apavlenko** [comment]
@alalek, could you please review?

File: modules/imgproc/src/canny.cpp [hunk 4]
```diff
@@ -805,6 +865,20 @@ void Canny( InputArray _src, OutputArray _dst,
 
     Mat src = _src.getMat(), dst = _dst.getMat();
 
+    CV_OVX_RUN(
+        false && /* disabling due to accuracy issues */
```
Comment (alalek) [inline modules/imgproc/src/canny.cpp]: Did you validate Canny test results?

Comment (apavlenko) [inline modules/imgproc/src/canny.cpp]: Sure! It fails tests for Canny accuracy, perf+sanity and even HoughLines...

Comment (alalek) [inline modules/imgproc/src/canny.cpp]: > It fails

Why?

Comment (apavlenko) [inline modules/imgproc/src/canny.cpp]: Look at the following example:
(source, canny-cv-canny-vx-diff)
canny-img-cv-vx-diff

Comment (alalek) [inline modules/imgproc/src/canny.cpp]: Looks like to Khronos OpenVX sample implementation bug. Other OpenVX implementations may work better.
/cc @vpisarev

Comment (apavlenko) [inline modules/imgproc/src/canny.cpp]: But generally I agree with @alalek: the `Imgproc_Canny.accuracy` and  `Imgproc_Canny.accuracy_deriv` are **very bad tests**!
I'm going to disable them and add another one - comparing VX vs non-VX results...

**Comment by apavlenko** [comment]
@alalek, I removed VX-HAL moving commits from this PR - could you merge it as soon as it passes the pre-commit tests?

File: modules/imgproc/test/ocl/test_canny.cpp [hunk 5]
```diff
@@ -133,6 +133,92 @@ OCL_INSTANTIATE_TEST_CASE_P(ImgProc, Canny, testing::Combine(
                                 testing::Values(L2gradient(false), L2gradient(true)),
                                 testing::Values(UseRoi(false), UseRoi(true))));
 
-} } // namespace cvtest::ocl
+
+IMPLEMENT_PARAM_CLASS(ImagePath, string)
+//IMPLEMENT_PARAM_CLASS(ApertureSize, int)
+//IMPLEMENT_PARAM_CLASS(L2gradient, bool)
+
+PARAM_TEST_CASE(CannyVX, ImagePath, ApertureSize, L2gradient)
```
Comment (alalek) [inline modules/imgproc/test/ocl/test_canny.cpp]: This test is not related to OpenCL.

Comment (apavlenko) [inline modules/imgproc/test/ocl/test_canny.cpp]: I know, but I reused some stuff from there

File: modules/imgproc/test/ocl/test_canny.cpp [hunk 6]
```diff
@@ -133,6 +133,92 @@ OCL_INSTANTIATE_TEST_CASE_P(ImgProc, Canny, testing::Combine(
...
+
+        setUseOpenVX(false);
+        Mat diff, diff1;
+        absdiff(canny, cannyVX, diff);
+        boxFilter(diff, diff1, -1, Size(3,3));
+        diff1 = diff1 > 255/9*3;
+        erode(diff1, diff1, Mat());
+        double error = cv::norm(diff1, NORM_L1) / 255;
+        const int maxError = 10;
+        if(error > maxError)
```
Comment (alalek) [inline modules/imgproc/test/ocl/test_canny.cpp]: 1) `cvtest::norm()` should be used instead of second `setUseOpenVX(false);`
2) this check is very strange and it is not clear. Why 10? What is about **relative** L1 norm?

**Comment by alalek** [comment]
Could you please extract changes into separate PR from these common files:
- modules/core/include/opencv2/core/openvx/ovx_defs.hpp
- modules/core/include/opencv2/core/ovx.hpp
- modules/core/src/ovx.cpp
- and related changes from 3rdparty/openvx/include/ivx.hpp

File: modules/imgproc/test/test_canny.cpp [hunk 7]
```diff
@@ -302,7 +302,8 @@ int CV_CannyTest::validate_test_results( int test_case_idx )
     return code;
 }
 
-TEST(Imgproc_Canny, accuracy) { CV_CannyTest test; test.safe_run(); }
-TEST(Imgproc_Canny, accuracy_deriv) { CV_CannyTest test(true); test.safe_run(); }
+// disabling, since testing on a white noise seems having too few sense...
+TEST(DISABLED_Imgproc_Canny, accuracy) { CV_CannyTest test; test.safe_run(); }
+TEST(DISABLED_Imgproc_Canny, accuracy_deriv) { CV_CannyTest test(true); test.safe_run(); }
```
Comment (alalek) [inline modules/imgproc/test/test_canny.cpp]: So my suggestion is to replace white noise to image list + adequate for there images parameters

File: modules/imgproc/test/ocl/test_canny.cpp (hunk 6 above)
Comment (apavlenko) [inline modules/imgproc/test/ocl/test_canny.cpp]: if you lose a line, it's not important how large the image is!
I don't think that relative norm is appropriate here...

Comment (alalek) [inline modules/imgproc/test/ocl/test_canny.cpp]: > if you lose a line

"Two pixel" lines on an "edge" of threshold should not be counted.

> it's not important how large the image is!

Relative to image of "lines" (reference) - IMHO it is a natural criteria, that should work.

**Comment by apavlenko** [comment]
@alalek ?

**Comment by alalek** [comment]
:+1:
